tests/test_models.py

```diff
@@ -1,30 +1,75 @@
 #!/usr/bin/env python3
 import plenoptic as po
+import pytest
+import matplotlib.pyplot as plt
+import torch
+
+# import plenoptic.simulate.models.naive
+import plenoptic
+from plenoptic.simulate.canonical_computations import gaussian1d, circular_gaussian2d
 from conftest import DEVICE
 
 
-class TestLinear(object):
+@pytest.fixture()
+def image_input():
+    return torch.rand(1, 1, 100, 100)
 
-    def test_linear(self, basic_stim):
-        model = po.simul.Linear().to(DEVICE)
-        assert model(basic_stim).requires_grad
 
-    def test_linear_metamer(self, einstein_img):
-        model = po.simul.Linear().to(DEVICE)
-        M = po.synth.Metamer(einstein_img, model)
-        M.synthesize(max_iter=3, learning_rate=1, seed=1)
+class TestFrontEnd:
+
+    all_models = [
+        "frontend.LinearNonlinear",
+        "frontend.LuminanceGainControl",
+        "frontend.LuminanceContrastGainControl",
+        "frontend.OnOff",
+    ]
+
+    @pytest.mark.parametrize("model", all_models[:-1], indirect=True)
+    def test_output_shape(self, model):
+        img = torch.ones(1, 1, 100, 100).to(DEVICE)
+        assert model(img).shape == img.shape
+
+    @pytest.mark.parametrize("model", all_models, indirect=True)
+    def test_gradient_flow(self, model):
+        img = torch.ones(1, 1, 100, 100).to(DEVICE)
+        y = model(img)
+        assert y.requires_grad
+
+    def test_onoff(self):
+        mdl = po.simul.OnOff(7, pretrained=False).to(DEVICE)
 
+    def test_pretrained_onoff(self):
+        mdl = po.simul.OnOff(7, pretrained=True).to(DEVICE)
```

Comment:
this is a weird test -- `kernel_size=7` is ignored if `pretrained=True` (see comment above)

Reply:
Fixed and added new test that tests for assertionwarning when kernel_size does not match 31x31